### PR TITLE
8326949: Authorization header is removed when a proxy Authenticator is set on HttpClient

### PR DESCRIPTION
**Pull request by Michael-Mc-Mahon**
This fix relaxes the constraints on user set authentication headers. Currently, any user set authentication headers are filtered out, if the HttpClient has an Authenticator set. The reason being that the authenticator is expected to manage authentication.  With this fix, it will be possible to use pre-emptive authentication through user set headers, even if an authenticator is set. The expected use case is where the authenticator would manage either proxy or server authentication and the user set headers would manage server authentication if the authenticator is managing proxy (or vice versa).
If the pre-emptive authentication fails, then this behavior is disabled on further retries and it would be up to the authenticator to provide the right credentials then.

Thanks,
Michael

---------
### Progress
- [ ] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8326949](https://bugs.openjdk.org/browse/JDK-8326949): Authorization header is removed when a proxy Authenticator is set on HttpClient (**Enhancement** - P3)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/21249/head:pull/21249` \
`$ git checkout pull/21249`

Update a local copy of the PR: \
`$ git checkout pull/21249` \
`$ git pull https://git.openjdk.org/jdk.git pull/21249/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 21249`

View PR using the GUI difftool: \
`$ git pr show -t 21249`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/21249.diff">https://git.openjdk.org/jdk/pull/21249.diff</a>

</details>

### Webrev
[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/21249#issuecomment-2381424613)

### CODE REVIEW

**Comment by AlanBateman** [comment]
Would there be any merit to adding something to Builder.authenticator(Authenticator) to document the behavior? I don't think know if this would be an API note or impNote. I'm mostly wondering where a developer might read about how/when it is used.

File: test/jdk/java/net/httpclient/UserAuthWithAuthenticator.java [hunk 1]
```diff
@@ -0,0 +1,318 @@
...
+            }
+            return new PasswordAuthentication("proxyUser", "proxyPwd".toCharArray());
+        }
+    }
+
+    static class ServerAuth extends Authenticator {
+        @Override
+        protected PasswordAuthentication getPasswordAuthentication() {
+            if (getRequestorType() != RequestorType.SERVER) {
+                // We only want to handle proxy authentication here
```
Comment (dfuch) [inline test/jdk/java/net/httpclient/UserAuthWithAuthenticator.java:282]:
```suggestion
                // We only want to handle server authentication here
```

File: src/java.net.http/share/classes/jdk/internal/net/http/common/Utils.java [hunk 2]
```diff
@@ -217,17 +217,39 @@ private static Set<String> getDisallowedHeaders() {
...
+        return (k, v) -> client.authenticator().isEmpty()
+            || (!k.equalsIgnoreCase("Authorization")
+                && !k.equalsIgnoreCase("Proxy-Authorization"))
+
+            // flag may be true for first attempt, and will be false
+            // for subsequent attempts if the first attempt failed
+            // due to 401/407
+
+            || req.tryUserSetAuthorization();
+*/
```
Comment (dfuch) [inline src/java.net.http/share/classes/jdk/internal/net/http/common/Utils.java:233]: Shouldn't this commented code be removed?

File: test/jdk/java/net/httpclient/UserAuthWithAuthenticator.java [hunk 3]
```diff
@@ -0,0 +1,318 @@
...
+        public void stopMocker() {
+            close(ss, s, in, out);
+        }
+
+        public int getPort() {
+            return ss.getLocalPort();
+        }
+
+        public String baseURL() {
+            return "http://127.0.0.1:" + getPort();
```
Comment (dfuch) [inline test/jdk/java/net/httpclient/UserAuthWithAuthenticator.java:212]: Should use `URIBuilder` here to deal with possible IPv6-only envs

**Comment by Michael-Mc-Mahon** [comment]
> Would there be any merit to adding something to Builder.authenticator(Authenticator) to document the behavior? I don't know if this would be an API note or impNote. I'm mostly wondering where a developer might read about how/when it is used.

That would be useful. It definitely should be documented somewhere.

**Comment by Michael-Mc-Mahon** [comment]
Quite a few existing authentication regression tests are failing with the change. So, I need to investigate them and will address the other PR comments after that.

**Comment by djelinski** [comment]
Would it make sense to assume that the user is always right, and use the authenticator only for the credentials that the user didn't provide?

I realize that it would be a major behavior change, but I think that's the least surprising behavior.

**Comment by Michael-Mc-Mahon** [comment]
> Would it make sense to assume that the user is always right, and use the authenticator only for the credentials that the user didn't provide?
> 
> I realize that it would be a major behavior change, but I think that's the least surprising behavior.

That could be simpler and still fits what I think the bug reporters are looking for which is basically that the authenticator would be used for one of server or proxy, while the user would (directly) look after the other.

The problem is partly caused by the fact that there's no way to register an authenticator for proxy only or server only or two separate authenticators for each.

**Comment by dfuch** [comment]
> >     Would it make sense to assume that the user is always right, and use the authenticator only for the credentials that the user didn't provide?
> > 
> >     I realize that it would be a major behavior change, but I think that's the least surprising behavior.
> > 
> 
> That could be simpler and still fits what I think the bug reporters are looking for which is basically that the authenticator would be used for one of server or proxy, while the user would (directly) look after the other.
> 
> The problem is partly caused by the fact that there's no way to register an authenticator for proxy only or server only or two separate authenticators for each.

I guess we could arrange to call the authenticator for `WWW-Authenticate` only if the user headers do not contain `Authorization` - and likewise for `Proxy-Authenticate` / `Proxy-Authorization`. 
If user headers contain Authorization and we receive 401, we just relay the 401 to the user.
Same for Proxy-Authorization / 407 ?

That would definitely need a CSR + release note if we changed the behavior that way though...

**Comment by Michael-Mc-Mahon** [comment]
> > > ```
> > > Would it make sense to assume that the user is always right, and use the authenticator only for the credentials that the user didn't provide?
> > > 
> > > I realize that it would be a major behavior change, but I think that's the least surprising behavior.
> > > ```
> > 
> > 
> > That could be simpler and still fits what I think the bug reporters are looking for which is basically that the authenticator would be used for one of server or proxy, while the user would (directly) look after the other.
> > The problem is partly caused by the fact that there's no way to register an authenticator for proxy only or server only or two separate authenticators for each.
> 
> I guess we could arrange to call the authenticator for `WWW-Authenticate` only if the user headers do not contain `Authorization` - and likewise for `Proxy-Authenticate` / `Proxy-Authorization`. If user headers contain Authorization and we receive 401, we just relay the 401 to the user. Same for Proxy-Authorization / 407 ?
> 
> That would definitely need a CSR + release note if we changed the behavior that way though...

I was thinking this needed a CSR and release note anyway.

**Comment by dfuch** [comment]
> I was thinking this needed a CSR and release note anyway.
true - but the risk of causing regression in existing code might be higher?

**Comment by Michael-Mc-Mahon** [comment]
> > I was thinking this needed a CSR and release note anyway.
> 
> true - but the risk of causing regression in existing code might be higher?

Maybe, but it seems unlikely that existing code would set an authenticator object and also set an authentication header which ends up not being used. Any code that sets an authentication header is probably using valid credentials I would think.

**Comment by dfuch** [comment]
Just to be clear I'm not objecting to implementing @djelinski suggestion. I believe it is a good suggestion which would be less surprising for first time users of that 'feature'.

**Comment by Michael-Mc-Mahon** [comment]
I just pushed an updated implementation, but still have to address some review comments. The next push should deal with remaining comments.

File: src/java.net.http/share/classes/java/net/http/HttpClient.java [hunk 4]
```diff
@@ -410,6 +410,11 @@ public interface Builder {
         /**
          * Sets an authenticator to use for HTTP authentication.
          *
+         * @apiNote
+         * If a {@link HttpRequest} has an {@code Authorization} or {@code
+         * Proxy-Authorization} header set then its value will override any
+         * value derived from the given {@link Authenticator}.
+         *
```
Comment (dfuch) [inline src/java.net.http/share/classes/java/net/http/HttpClient.java:417]: This sounds like we're going to call the authenticator, but we're not. Maybe:
```suggestion
         * @apiNote
         * If a {@link HttpRequest} has an {@code Authorization} or {@code
         * Proxy-Authorization} header set then its value will be used and
         * the {@link Authenticator} will not be invoked for the corresponding
         * authentication.
         *
```

**Comment by Michael-Mc-Mahon** [comment]
It looks like I made a mess of the last resync and rebased the latest changes from master onto my own branch. I will have to close this PR and create a new one with the current patch.

**Pull request closed without merging**